contracts/child/ChildToken/BaseERC20.sol

```diff
@@ -0,0 +1,357 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.6.6;
+
+/**
+ * @dev Implementation of the {IERC20} interface.
+ *
+ * This implementation is agnostic to the way tokens are created. This means
+ * that a supply mechanism has to be added in a derived contract using {_mint}.
+ * For a generic mechanism see {ERC20PresetMinterPauser}.
+ *
+ * TIP: For a detailed writeup see our guide
+ * https://forum.zeppelin.solutions/t/how-to-implement-erc20-supply-mechanisms/226[How
+ * to implement supply mechanisms].
+ *
+ * We have followed general OpenZeppelin guidelines: functions revert instead
+ * of returning `false` on failure. This behavior is nonetheless conventional
+ * and does not conflict with the expectations of ERC20 applications.
+ *
+ * Additionally, an {Approval} event is emitted on calls to {transferFrom}.
+ * This allows applications to reconstruct the allowance for all accounts just
+ * by listening to said events. Other implementations of the EIP may not emit
+ * these events, as it isn't required by the specification.
+ *
+ * Finally, the non-standard {decreaseAllowance} and {increaseAllowance}
+ * functions have been added to mitigate the well-known issues around setting
+ * allowances. See {IERC20-approve}.
+ */
+contract BaseERC20 {
+    mapping (address => uint256) private _balances;
+
+    mapping (address => mapping (address => uint256)) private _allowances;
+
+    uint256 private _totalSupply;
+
+    string private _name;
+    string private _symbol;
+    uint8 private _decimals;
+
+    /**
+     * @dev Emitted when `value` tokens are moved from one account (`from`) to
+     * another (`to`).
+     *
+     * Note that `value` may be zero.
+     */
+    event Transfer(address indexed from, address indexed to, uint256 value);
+
+    /**
+     * @dev Emitted when the allowance of a `spender` for an `owner` is set by
+     * a call to {approve}. `value` is the new allowance.
+     */
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+
+    /**
+     * @dev Emitted when you're interested in **actually** burning token in both L2, L1
+     *
+     * L1 Predicate contract expects this event signature, when you expect it to burn token
+     * on L1 too.
+     *
+     * NOTE L2 burning is done by invoking `withdraw` function defined in child token, which
+     * will eventually emit this event
+     */
+    event Burn(address indexed from, uint256 amount);
```

Comment:
I am not sure I follow the idea here.

ERC20 have already a burn event, this is the Transfer event where the to = address(0).
This should be considered as the burn case. 

a ChainExit event would be more logical in my opinion. It would be triggered when you want the token to be unlocked from L1.


`ChainExit(address to, uint256 amount)` could for example work

I would personally make it more generic like so:

`ChainExit(uint256 chainID, address to, uint256 amount)`


If no such event happen the token is basically burn on L1 as it get forever locked.

I can see though the benefit of a specifuc Burn even like you did to allow anyone to actually burn the token on L1. but I am not sure that in practise it is very usefl. Who is going to pay the gas cost to actually burn the tokens on L1. There are no incentives to do so.

Reply:
- `Burn` -> true burning on both chains
- `Transfer` -> Bring tokens from L2 to L1

`burn` is supposed to be invoked by ADMIN on L2.

Reply:
I don't think this is the right approach to the problem as I mentioned in the other comments

Reply:
I agree with Ronan here - we shouldn't use that many events when we can cover all the cases using one event.

Reply:
Burning on L1 makes 0 sense, nobody will actually "exit" such event. Burning should just happen on L2 and be done with that.